librz/bin/format/mach0/dyldcache.c

```diff
@@ -1,295 +1,1084 @@
+		}
+
+		result->entries = infos;
+		result->length = j;
+		return result;
+	}
+
+	if (cache->hdr->mappingCount > 1) {
+		RzDyldRebaseInfosEntry *infos = RZ_NEWS0(RzDyldRebaseInfosEntry, 1);
+		if (!infos) {
+			goto beach;
+		}
+
+		infos[0].start = cache->maps[1].fileOffset;
+		infos[0].end = infos[0].start + cache->maps[1].size;
+		infos[0].info = get_rebase_info(cache, cache->hdr->slideInfoOffset, cache->hdr->slideInfoSize, infos[0].start, UT64_MAX);
+
+		result->entries = infos;
+		result->length = 1;
+		return result;
+	}
+
+beach:
+	free(result);
 	return NULL;
 }
 
-void rz_bin_dydlcache_get_libname(struct rz_bin_dyldcache_lib_t *lib, char **libname) {
-	char *cur = lib->path;
-	char *res = lib->path;
-	int path_length = strlen(lib->path);
-	while (cur < cur + path_length - 1) {
-		cur = strchr(cur, '/');
-		if (!cur) {
-			break;
+RZ_API ut64 rz_dyldcache_get_slide(RzDyldCache *cache) {
+	rz_return_val_if_fail(cache, 0);
+	if (!cache->rebase_infos || !cache->rebase_infos->length) {
+		return 0;
+	}
+
+	size_t i;
+	for (i = 0; i < cache->rebase_infos->length; i++) {
+		if (cache->rebase_infos->entries[i].info) {
+			return cache->rebase_infos->entries[i].info->slide;
 		}
-		cur++;
-		res = cur;
 	}
-	*libname = res;
+
+	return 0;
 }
 
-struct rz_bin_dyldcache_obj_t *rz_bin_dyldcache_new(const char *file) {
-	struct rz_bin_dyldcache_obj_t *bin;
-	if (!(bin = RZ_NEW0(struct rz_bin_dyldcache_obj_t))) {
+RZ_API RzDyldCache *rz_dyldcache_new_buf(RzBuffer *buf) {
+	RzDyldCache *cache = RZ_NEW0(RzDyldCache);
+	if (!cache) {
 		return NULL;
 	}
-	bin->file = file;
-	size_t binsz;
-	ut8 *buf = (ut8 *)rz_file_slurp(file, &binsz);
-	bin->size = binsz;
-	if (!buf) {
-		return rz_bin_dyldcache_free(bin);
+	memcpy(cache->magic, "dyldcac", 7);
+	cache->buf = rz_buf_ref(buf);
+	cache->hdr = read_cache_header(cache->buf);
+	if (!cache->hdr) {
+		goto cupertino;
 	}
-	bin->b = rz_buf_new_with_bytes(NULL, 0);
-	if (!rz_buf_set_bytes(bin->b, buf, bin->size)) {
-		free(buf);
-		return rz_bin_dyldcache_free(bin);
+	cache->maps = read_cache_maps(cache->buf, cache->hdr);
+	if (!cache->maps) {
+		goto cupertino;
 	}
-	free(buf);
-	if (!rz_bin_dyldcache_init(bin)) {
-		return rz_bin_dyldcache_free(bin);
+	cache->accel = read_cache_accel(cache->buf, cache->hdr, cache->maps);
+	cache->locsym = rz_dyld_locsym_new(cache->buf, cache->hdr);
+	cache->bins = create_cache_bins(cache->buf, cache->hdr, cache->maps, cache->accel);
+	if (!cache->bins) {
+		goto cupertino;
 	}
-	return bin;
+	cache->rebase_infos = get_rebase_infos(cache);
+	return cache;
+cupertino:
+	rz_dyldcache_free(cache);
+	return NULL;
 }
 
-struct rz_bin_dyldcache_obj_t *rz_bin_dyldcache_from_bytes_new(const ut8 *buf, ut64 size) {
-	struct rz_bin_dyldcache_obj_t *bin = RZ_NEW0(struct rz_bin_dyldcache_obj_t);
-	if (!bin) {
+static void rebase_info3_free(RzDyldRebaseInfo3 *rebase_info) {
+	if (!rebase_info) {
+		return;
+	}
+
+	RZ_FREE(rebase_info->page_starts);
+	RZ_FREE(rebase_info);
```

Comment:
Same here - will not work for `rebase_info` since it's an argument of the function. Do we really need `RZ_FREE()` macro here?